File: datamodel/src/main/java/datamodel/clinical/treatment/history/TreatmentResponse.java
package datamodel.clinical.treatment.history;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public enum TreatmentResponse {
    PROGRESSIVE_DISEASE,
    STABLE_DISEASE,
    MIXED,
    PARTIAL_RESPONSE,
    COMPLETE_RESPONSE,
    REMISSION;

    @Nullable
    public static TreatmentResponse createFromString(@NotNull String input) {
        switch (input.toUpperCase()) {
            case "PD":
                return PROGRESSIVE_DISEASE;
            case "SD":
                return STABLE_DISEASE;
            case "MIXED":
                return MIXED;
            case "PR":
                return PARTIAL_RESPONSE;
            case "CR":  // not seen in existing curation
                return COMPLETE_RESPONSE;
            case "REMISSION":
                return REMISSION;
            default:
                return null;
        }
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/datamodel/clinical/treatment/history/TreatmentResponse.kt
package com.hartwig.actin.datamodel.clinical.treatment.history

import com.hartwig.actin.datamodel.Displayable

enum class TreatmentResponse : Displayable {
    PROGRESSIVE_DISEASE,
    STABLE_DISEASE,
    MIXED,
    PARTIAL_RESPONSE,
    NEAR_COMPLETE_RESPONSE,
    COMPLETE_RESPONSE,
    REMISSION;

    override fun display(): String {
        return this.toString().replace("_", " ").lowercase()
    }

    companion object {
        val BENEFIT_RESPONSES = setOf(PARTIAL_RESPONSE, NEAR_COMPLETE_RESPONSE, COMPLETE_RESPONSE, REMISSION)

        fun createFromString(input: String): TreatmentResponse? {
            return when (input.uppercase()) {
                "PD" -> PROGRESSIVE_DISEASE
                "SD" -> STABLE_DISEASE
                "MIXED" -> MIXED
                "PR" -> PARTIAL_RESPONSE
                "CR" -> COMPLETE_RESPONSE
                "REMISSION" -> REMISSION
                else -> null
            }
        }
    }
}
